src/bpf/bpfctl.c

```diff
@@ -0,0 +1,238 @@
+        uint64_t type;
+        uint64_t memlock;
+};
+
+static void bpf_prog_data_done(BpfProgramData *p) {
+        assert(p);
+
+        p->name = mfree(p->name);
+        p->tag = mfree(p->tag);
+        strv_free(p->map_names);
+}
+
+static void bpf_map_data_done(BpfMapData *p) {
+        assert(p);
+
+        p->name = mfree(p->name);
+}
+
+static int query_bpf_progs_and_maps(void) {
+        _cleanup_(varlink_unrefp) Varlink *vl = NULL;
+        _cleanup_(table_unrefp) Table *map_table = NULL, *prog_table = NULL;
+        JsonVariant *i = NULL, *o = NULL, *m = NULL, *v = NULL;
+        int r;
+
+        map_table = table_new("map_name", "map_type", "memlock");
+        if (!map_table)
+                return log_oom();
+
+        prog_table = table_new("program_name", "program_type", "memlock", "program_tag", "program_map_names", "run_time_ns", "run_cnt");
+        if (!prog_table)
+                return log_oom();
+
+        r = varlink_connect_address(&vl, "/run/systemd/bpf/io.systemd.Bpf");
+        if (r < 0)
+                return log_error_errno(r, "Unable to connect to /run/systemd/bpf/io.systemd.Bpf: %m");
+
+        r = varlink_collect_full(vl, "io.systemd.Bpf.Describe", i, &o, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to collect varlink method: %m");
+
+        JSON_VARIANT_ARRAY_FOREACH(i, o) {
+                m = json_variant_by_key(i, "type");
+                if (!m)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to find type field of json variant");
+                v = json_variant_by_key(i, "data");
+                if (!v)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to find data field of json variant");
+
+                if (streq(json_variant_string(m), "map")) {
+                        static const JsonDispatch dispatch_table[] = {
+                                { "name",  JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct BpfMapData, name),  JSON_MANDATORY },
+                                { "type", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfMapData, type), JSON_MANDATORY},
+                                { "memlock", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfMapData, memlock), JSON_MANDATORY},
+                                {}
+                        };
+                        _cleanup_(bpf_map_data_done) BpfMapData p = {};
+
+                        r = json_dispatch(v, dispatch_table, 0, &p);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to dispatch map table: %m");
+
+                        r = table_add_many(map_table,
+                                   TABLE_STRING, p.name, TABLE_SET_ALIGN_PERCENT, 100,
+                                   TABLE_UINT64, p.type, TABLE_SET_ALIGN_PERCENT, 100,
+                                   TABLE_UINT64, p.memlock, TABLE_SET_ALIGN_PERCENT, 100);
```

Comment:
hmm, if "memlock" refers to bytes of memory, then should use TABLE_SIZE as field type